### ISSUE TITLE
oauth2-bundle as server with HWIOAuthBundle as client

### ISSUE THREAD

**Issue by nsc-novacens** [trikoder/oauth2-bundle#218]
Hello,

I'm using HWIOAuthBundle  with custom oAuth2 (And FaceBook / Google) in Symfony 3.4 or 5.1 project.

Now i'm trying to use trikoder/oauth2-bundle on a Symfony 3.4 project (Web Application with FOSUser) to autorize acces from remote users.

But it's not working ...

I can't pass though the "/api/user" with curl, always getting "The resource server rejected the request.", /api/token is working it give me a access_token and refresh_token

With HWIOAuthBundle i get : {"error":"invalid_client","error_description":"Client authentication failed","message":"Client authentication failed"}

Error 401 :/

Thanks

Nicolas

**Comment by X-Coder264** [comment]
Please paste the exact cURL request and any other data (like how the client entry looks in the database) that we can use to help you with this issue.

The message `Client authentication failed` gets shown when `OAuthServerException::invalidClient()` gets thrown.

https://github.com/thephpleague/oauth2-server/blob/8.1.1/src/Grant/AbstractGrant.php#L178 -> this is the place where you should look.

**Comment by nsc-novacens** [comment]
Hello,

It took me most of the night, but it works!

I started from scratch.

1. Create a "password" client and debug my implementation of **api/token** and **api/user-info** with CURL.
2. Create an "authorization_code" client, read [PR177](https://github.com/trikoder/oauth2-bundle/blob/ead963bcba0a3436fde7395c6957134c1cdcc4e7/docs/authorization-code-grant.md), adapt the code for Symfony 3.4 and oauth2-bundle v2.0
3. Create a custom oAuth in HWIOAuthBundle (ResourceOwner) and adapt FOSUBUserPRovider to handle this new ResourceOwner.
4. Drink coffee, dump() and die() step by step.

One of the big worries was to think about taking the "authorize" route out of the api routes otherwise users are authenticated on the wrong Symfony firewall.

I also lost time with the 'Client authentication failed' error, it came from the redirect_uri which were different.

Luckily I've already implemented two custom oAuth Client with HWIOAuthBundle (From [GenericOAuth2ResourceOwner](https://github.com/hwi/HWIOAuthBundle/blob/master/OAuth/ResourceOwner/GenericOAuth2ResourceOwner.php)) it helped me a lot, without it I think it's almost impossible...

You should validate this PR177 !

Thanks

Nicolas

**Comment by nsc-novacens** [comment]
Everything works fine, I was able to test with another site the complete flow.

1. Site www.client-hwioauth.com added a link to log in with an account from www.server-trikoder.com (like a FaceBook / Google)
2. If the person is not logged in www.server-trikoder.com, he or she arrives at www.server-trikoder.com with an authentication page (email / password). If the connection is successful, they are redirected to www.client-hwioauth.com and are "connected".
3. If the person is already connected to www.server-trikoder.com, he or she is automatically "connected" (double redirection).

**Comment by X-Coder264** [comment]
I'm glad that you got it working. Since the issue is solved, I'm closing it.

**Comment by mnsami** [comment]
hey @nsc-novacens @X-Coder264 

I'm trying to do the same as you (i think) using `authorization_grant`, but I need some help.

Here what I have:
1. A symfony 4.4 API project using FOSRestBundle - API first design that has all the resources I want to protect
2. I installed `nucleos/user-bundle` for user management.
3. I have installed `trikoder/oauth2` and hooked it up with `nucleos/user-bundle` (all installed in the same Symfony 4.4 project)
4. I have setup a listener to listen to `AuthorizationRequestResolveEvent` and set the user successfully.
5. I have a ReactJS frontend which I want my users to login from and access my resources.

What I want to do is:
1. Login via API, and on successfully login, the user is authorized behind the scenes and an access token is returned to the FE (reactjs)
2. Using the access token the reactJS will do any further API requests.

Is what i'm trying to do logical ?
